gk/main.c

```diff
@@ -850,6 +850,101 @@ add_ggu_policy(struct ggu_policy *policy,
 	}
 }
 
+static void
+flush_flow_table(struct ip_prefix *src,
+	struct ip_prefix *dst, struct gk_instance *instance)
+{
+	uint16_t proto = 0;
+	uint32_t next = 0;
+	int32_t index;
+	const struct ip_flow *key;
+	void *data;
+	struct in_addr ip4_src_mask;
+	struct in_addr ip4_dst_mask;
+	struct in6_addr ip6_src_mask;
+	struct in6_addr ip6_dst_mask;
+
+	memset(&ip4_src_mask, 0, sizeof(ip4_src_mask));
+	memset(&ip4_dst_mask, 0, sizeof(ip4_dst_mask));
+	memset(&ip6_src_mask, 0, sizeof(ip6_src_mask));
+	memset(&ip6_dst_mask, 0, sizeof(ip6_dst_mask));
+
+	if (src != NULL) {
+		if (src->addr.proto == ETHER_TYPE_IPv4) {
+			ip4_prefix_mask(src->len, &ip4_src_mask);
+			proto = ETHER_TYPE_IPv4;
+		} else {
+			ip6_prefix_mask(src->len, &ip6_src_mask);
+			proto = ETHER_TYPE_IPv6;
+		}
+	}
+
+	if (dst != NULL) {
+		if (dst->addr.proto == ETHER_TYPE_IPv4) {
+			ip4_prefix_mask(dst->len, &ip4_dst_mask);
+			proto = ETHER_TYPE_IPv4;
+		} else {
+			ip6_prefix_mask(dst->len, &ip6_dst_mask);
+			proto = ETHER_TYPE_IPv6;
+		}
+	}
+
+	index = rte_hash_iterate(instance->ip_flow_hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		bool matched = true;
+		struct flow_entry *fe =
+			&instance->ip_flow_entry_table[index];
+
+		if (proto != fe->flow.proto)
+			goto next;
+
+		if (proto == ETHER_TYPE_IPv4) {
+			if (src != NULL) {
+				matched &= ip4_same_subnet(
+					src->addr.ip.v4.s_addr,
+					fe->flow.f.v4.src,
+					ip4_src_mask.s_addr);
+			}
+
+			if (dst != NULL) {
+				matched &= ip4_same_subnet(
+					dst->addr.ip.v4.s_addr,
+					fe->flow.f.v4.dst,
+					ip4_dst_mask.s_addr);
+			}
+		}
+
+		if (proto == ETHER_TYPE_IPv6) {
+			if (src != NULL) {
+				matched &= ip6_same_subnet(
+					src->addr.ip.v6.s6_addr,
+					fe->flow.f.v6.src,
+					ip6_src_mask.s6_addr);
+			}
+
+			if (dst != NULL) {
+				matched &= ip6_same_subnet(
+					dst->addr.ip.v6.s6_addr,
+					fe->flow.f.v6.dst,
+					ip6_dst_mask.s6_addr);
+			}
+		}
+
+		if (matched)
+			gk_del_flow_entry_from_hash(
+				instance->ip_flow_hash_table, fe);
+
+next:
+		index = rte_hash_iterate(instance->ip_flow_hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	GK_LOG(NOTICE,
+		"The GK block finished flushing flow table at %s with lcore %u\n",
+		__func__, rte_lcore_id());
```

Comment:
This log entry would be more helpful if it included the number of flow entries removed.